BMI string routines assign non-conforming arrays from string_to_char_array

> [!NOTE]
> Confirmed under intel-classic and fixed by #2902. The overrun does not occur with gfortran, which sizes the copy by the function result; see https://github.com/MODFLOW-ORG/modflow6/issues/2901#issuecomment-5133354735.

### Summary

Three BMI routines assign the result of `string_to_char_array` to a dummy argument with a different extent. The function result has `length + 1` elements, while the dummy argument is declared with the corresponding BMI length constant. The assignment is not conforming, and in practice the shorter temporary is read past its end.

| routine | file | dummy extent | result extent |
| --- | --- | --- | --- |
| `get_version` | `srcbmi/mf6xmi.F90:353` | `BMI_LENVERSION` (256) | 11 |
| `get_component_name` | `srcbmi/mf6bmi.f90:51` | `BMI_LENCOMPONENTNAME` (256) | 10 |
| `get_grid_type` | `srcbmi/mf6bmiGrid.f90:84` | `BMI_LENGRIDTYPE` (17) | 12 or 13 |

```fortran
character(kind=c_char), intent(inout) :: mf_version(BMI_LENVERSION)
...
mf_version = string_to_char_array(VERSIONNUMBER, len_trim(VERSIONNUMBER))
```

Two other routines perform the same conversion with an explicit slice on the left-hand side, which conforms and is not affected:

- `get_var_address` (`srcbmi/mf6xmi.F90:407`)
- `get_var_type` (`srcbmi/mf6bmi.f90:1175`)

### How it surfaces

`string_to_char_array` writes the null terminator in the correct position, so the returned string is correct whenever the overread does not fault. The defect is therefore invisible in most builds and only appears when the read reaches unmapped memory, or when bounds checking is enabled.

We see it as an intermittent access violation on Windows in the mf6adj continuous integration, which drives MODFLOW 6 through modflowapi and the nightly build. Every Windows failure faults inside `get_version`:

```
adj.py:1036: in _get_gwf_version
    version = self._gwf.get_version()
xmiwrapper.py:237: in get_version
    self._execute_function(self.lib.get_version, byref(version))
E   OSError: exception: access violation reading 0x000002366A4D7000
```

Eight scheduled runs failed this way between 2026-05-18 and 2026-07-30, roughly one run in eight. The Linux and macOS jobs have never failed this way, and the Windows and Linux nightly binaries are built with the same compiler (intel-classic 2021.6), so this is a memory layout difference rather than a code generation difference. Representative runs:

- https://github.com/INTERA-Inc/mf6adj/actions/runs/30528548075
- https://github.com/INTERA-Inc/mf6adj/actions/runs/30252492007
- https://github.com/INTERA-Inc/mf6adj/actions/runs/30080410850
- https://github.com/INTERA-Inc/mf6adj/actions/runs/29401997112

The same defect was reported in #1993 as a gfortran runtime error on ARM mac:

> Dimension 1 of array 'c_array' has extent X instead of Y

That issue was resolved by converting the assignments to subroutine calls. #1996 reverted the subroutine and changed the dummy arguments from `intent(out)` to `intent(inout)`, which stops gfortran from reporting the mismatch but leaves the non-conforming assignment in place.

### Suggested fix

Slice the left-hand side, as `get_var_address` and `get_var_type` already do:

```fortran
mf_version(1:len_trim(VERSIONNUMBER) + 1) = &
  string_to_char_array(VERSIONNUMBER, len_trim(VERSIONNUMBER))
```

A develop profile build (`-Ddebug=true -Doptimization=0`) compiles with `-fcheck=all` and turns the overread into a hard runtime error, so an API test that calls these three routines will hold the fix in place.

Confirmed under intel-classic. The correction above applies to gfortran only.

I temporarily reverted the three assignments on the #2902 branch, keeping the new `autotest/test_libmf6_strings.py`, and let CI run both ways. The test fills the caller's buffer with a sentinel byte, calls the routine, and checks that nothing past the null terminator was written.

| | gfortran | intel-classic 2021.6 (windows-2022) |
| --- | --- | --- |
| with the fix | pass | pass |
| without the fix | pass | **fail** |

Without the fix, `Test Intel fortran (windows-2022)` reports one failure out of 802 tests:

```
File "autotest/test_libmf6_strings.py", line 175, in api_func
  assert call_string_routine(mf6, "get_version", "BMI_LENVERSION") == version
File "autotest/test_libmf6_strings.py", line 136, in call_string_routine
  assert tail == sentinel * len(tail), f"{name} wrote past the null terminator"
AssertionError: get_version wrote past the null terminator
assert b'\x00  s\x00...0\x00\x00\x00' == b'\xaa\xaa\xa...a\xaa\xaa\xaa'
```

intel-classic sizes the copy by the extent of the dummy argument rather than the extent of the function result, so `get_version` writes 256 elements and reads 245 elements past the end of an 11-element temporary. gfortran sizes the copy by the function result, which is why the defect is invisible there and why the earlier probe found nothing.

The read overrun is the fault mf6adj sees on Windows, where the release and nightly binaries are built with the same compiler. The mechanism in the original description holds; only the claim that it applied to every compiler was wrong.

Fixed by #2902, which also adds the test. `Test Intel fortran (windows-2022)` runs it on every pull request, so the overrun is caught deterministically rather than as an intermittent crash.

--- a/autotest/test_libmf6_strings.py
+++ b/autotest/test_libmf6_strings.py
@@ -1,0 +1,212 @@
+"""
+Test the BMI routines that return a C character string with a basic 1x1x5 model:
+
+  [ BC | 1 | 2 | 3 | BC ]
+
+Each routine fills a caller-allocated buffer sized by the matching BMI length
+constant, and must write no more than the string and its null terminator.
+
+Cases:
+  - libmf6_strings : get_component_name, get_version, and get_grid_type return
+                     the expected strings and leave the rest of the buffer alone.
+"""
+
+import os
+import re
+from ctypes import byref, c_int, create_string_buffer
+
+import flopy
+import pytest
+from framework import TestFramework
+from modflow_devtools.markers import requires_pkg
+
+cases = ["libmf6_strings"]
+model_name = "model"
+
+# byte used to fill the buffer before each call
+sentinel = b"\xaa"
+
+
+def get_model(dir):
+    # parameters and spd
+    # tdis
+    nper = 1
+    tdis_rc = []
+    for i in range(nper):
+        tdis_rc.append((1.0, 1, 1))
+
+    # solver data
+    nouter, ninner = 100, 300
+    hclose, rclose = 10e-9, 1e-3
+
+    # model spatial discretization
+    nlay = 1
+    nrow = 1
+    ncol = 5
+
+    # cell spacing
+    delr = 1.0
+    delc = 1.0
+
+    # top/bot of the aquifer
+    tops = [0.0, -1.0]
+
+    # hydraulic conductivity
+    k11 = 1.0
+
+    # boundary stress period data
+    h_left = 0.0
+    h_right = 5.0
+
+    # initial head
+    h_start = 0.0
+
+    sim = flopy.mf6.MFSimulation(
+        sim_name="sim", version="mf6", exe_name="mf6", sim_ws=dir
+    )
+
+    tdis = flopy.mf6.ModflowTdis(sim, time_units="DAYS", nper=nper, perioddata=tdis_rc)
+
+    ims = flopy.mf6.ModflowIms(
+        sim,
+        print_option="SUMMARY",
+        outer_dvclose=hclose,
+        outer_maximum=nouter,
+        inner_maximum=ninner,
+        inner_dvclose=hclose,
+        rcloserecord=rclose,
+        linear_acceleration="CG",
+    )
+
+    chd_data = [[(0, 0, 0), h_left], [(0, 0, ncol - 1), h_right]]
+    chd_spd = {0: chd_data}
+
+    gwf = flopy.mf6.ModflowGwf(sim, modelname=model_name, save_flows=True)
+    dis = flopy.mf6.ModflowGwfdis(
+        gwf,
+        nlay=nlay,
+        nrow=nrow,
+        ncol=ncol,
+        delr=delr,
+        delc=delc,
+        top=tops[0],
+        botm=tops[1:],
+    )
+    ic = flopy.mf6.ModflowGwfic(gwf, strt=h_start)
+    npf = flopy.mf6.ModflowGwfnpf(
+        gwf,
+        save_specific_discharge=True,
+        save_flows=True,
+        icelltype=0,
+        k=k11,
+    )
+    chd = flopy.mf6.ModflowGwfchd(gwf, stress_period_data=chd_spd)
+    oc = flopy.mf6.ModflowGwfoc(
+        gwf,
+        head_filerecord=f"{model_name}.hds",
+        budget_filerecord=f"{model_name}.cbc",
+        saverecord=[("HEAD", "LAST"), ("BUDGET", "LAST")],
+    )
+
+    return sim
+
+
+def build_models(idx, test):
+    # build MODFLOW 6 files
+    ws = test.workspace
+    sim = get_model(ws)
+
+    # build comparison model
+    ws = os.path.join(test.workspace, "libmf6")
+    sim_compare = get_model(ws)
+
+    return sim, sim_compare
+
+
+def call_string_routine(mf6, name, size_name, *args):
+    """Call a BMI string routine and return the string it wrote."""
+    size = mf6.get_constant_int(size_name)
+    buff = create_string_buffer(sentinel * size, size)
+    status = getattr(mf6.lib, name)(*args, byref(buff))
+    assert status == 0, f"{name} returned status {status}"
+
+    # everything past the null terminator must be untouched
+    value = buff.value.decode()
+    tail = buff.raw[len(value) + 1 :]
+    assert tail == sentinel * len(tail), f"{name} wrote past the null terminator"
+
+    return value
+
+
+def api_func(exe, idx, model_ws=None):
+    from modflowapi import ModflowApi
+
+    if model_ws is None:
+        model_ws = "."
+    output_file_path = os.path.join(model_ws, "mfsim.stdout")
+
+    try:
+        mf6 = ModflowApi(exe, working_directory=model_ws)
+    except Exception as e:
+        print("Failed to load " + str(exe))
+        print("with message: " + str(e))
+        return False, open(output_file_path).readlines()
+
+    # initialize the model
+    try:
+        mf6.initialize()
+    except:
+        return False, open(output_file_path).readlines()
+
+    grid_id = mf6.get_var_grid(mf6.get_var_address("X", model_name))
+
+    # strings returned through the wrapper
+    assert mf6.get_component_name() == "MODFLOW 6"
+    version = mf6.get_version()
+    assert re.match(r"^\d+\.\d+\.\d+", version), f"unexpected version '{version}'"
+    assert mf6.get_grid_type(grid_id) == "rectilinear"
+
+    # same strings, checking what each routine wrote into the buffer
+    assert (
+        call_string_routine(mf6, "get_component_name", "BMI_LENCOMPONENTNAME")
+        == "MODFLOW 6"
+    )
+    assert call_string_routine(mf6, "get_version", "BMI_LENVERSION") == version
+    assert (
+        call_string_routine(
+            mf6, "get_grid_type", "BMI_LENGRIDTYPE", byref(c_int(grid_id))
+        )
+        == "rectilinear"
+    )
+
+    # time loop
+    current_time = mf6.get_current_time()
+    end_time = mf6.get_end_time()
+    while current_time < end_time:
+        try:
+            mf6.update()
+        except:
+            return False, open(output_file_path).readlines()
+        current_time = mf6.get_current_time()
+
+    # finish
+    try:
+        mf6.finalize()
+    except:
+        return False, open(output_file_path).readlines()
+
+    # cleanup and return
+    return True, open(output_file_path).readlines()
+
+
+@requires_pkg("modflowapi")
+@pytest.mark.parametrize("idx, name", enumerate(cases))
+def test_mf6model(idx, name, function_tmpdir, targets):
+    test = TestFramework(
+        name=name,
+        workspace=function_tmpdir,
+        build=lambda t: build_models(idx, t),
+        targets=targets,
+        api_func=lambda exe, ws: api_func(exe, idx, ws),
+    )
+    test.run()

--- a/doc/ReleaseNotes/develop.toml
+++ b/doc/ReleaseNotes/develop.toml
@@ -184,3 +184,8 @@ description = "The Skeletal Storage, Compaction, and Subsidence (CSUB) Package p
 section = "fixes"
 subsection = "advanced"
 description = "The Skeletal Storage, Compaction, and Subsidence (CSUB) Package terminated when the summed interbed thickness in a cell exceeded the cell thickness, even when the difference was a numerical-roundoff artifact of the input precision. When the excess is within a small fraction (1e-5) of the cell thickness, the interbed thicknesses in the cell are now proportionally scaled to fit and the simulation continues. A larger excess is still an error."
+
+[[items]]
+section = "fixes"
+subsection = "basic"
+description = "The BMI get\\_version, get\\_component\\_name, and get\\_grid\\_type functions assigned the converted C string to a character array longer than the string itself. The assignment is not conforming, which left the amount of the caller's buffer that is written up to the compiler. The assignments were corrected to write only the string and its null terminator."

--- a/srcbmi/mf6bmi.f90
+++ b/srcbmi/mf6bmi.f90
@@ -47,8 +47,10 @@ contains
     character(kind=c_char), intent(inout) :: name(BMI_LENCOMPONENTNAME)
     integer(kind=c_int) :: bmi_status !< BMI status code
     ! -- local variables
+    character(len=*), parameter :: component_name = 'MODFLOW 6'
 
-    name = string_to_char_array('MODFLOW 6', 9)
+    name(1:len(component_name) + 1) = &
+      string_to_char_array(component_name, len(component_name))
     bmi_status = BMI_SUCCESS
 
   end function bmi_get_component_name

--- a/srcbmi/mf6bmiGrid.f90
+++ b/srcbmi/mf6bmiGrid.f90
@@ -81,7 +81,8 @@ contains
     else
       return
     end if
-    grid_type = string_to_char_array(trim(grid_type_f), len_trim(grid_type_f))
+    grid_type(1:len_trim(grid_type_f) + 1) = &
+      string_to_char_array(trim(grid_type_f), len_trim(grid_type_f))
     bmi_status = BMI_SUCCESS
   end function get_grid_type
 

--- a/srcbmi/mf6xmi.F90
+++ b/srcbmi/mf6xmi.F90
@@ -350,7 +350,8 @@ contains
     character(kind=c_char), intent(inout) :: mf_version(BMI_LENVERSION)
     integer(kind=c_int) :: bmi_status !< BMI status code
 
-    mf_version = string_to_char_array(VERSIONNUMBER, len_trim(VERSIONNUMBER))
+    mf_version(1:len_trim(VERSIONNUMBER) + 1) = &
+      string_to_char_array(VERSIONNUMBER, len_trim(VERSIONNUMBER))
     bmi_status = BMI_SUCCESS
 
   end function xmi_get_version

